flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -2113,6 +2122,63 @@ IntrinsicLibrary::genRepeat(mlir::Type resultType,
       });
 }
 
+// RESHAPE
+fir::ExtendedValue
+IntrinsicLibrary::genReshape(mlir::Type resultType,
+                             llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 4);
+
+  // Handle source argument
+  auto source = builder.createBox(loc, args[0]);
+
+  // Handle shape argument
+  auto shape = builder.createBox(loc, args[1]);
+  fir::BoxValue shapeTmp = shape;
+  auto shapeRank = shapeTmp.rank();
+  assert(shapeRank == 1);
+  auto shapeTy = shape.getType();
+  auto shapeArrTy = fir::dyn_cast_ptrOrBoxEleTy(shapeTy);
+  auto shapeSeqTy = shapeArrTy.cast<fir::SequenceType>();
+  auto shapeEleTy = shapeSeqTy.getEleTy();
+  assert(fir::isa_integer(shapeEleTy));
+  auto resultRank = shapeSeqTy.getShape();
```

Comment:
Grouping things into `auto resultRank = shapeArrTy.cast<fir::SequenceType>().getShape()[0];` would be easier to follow in my opinion.